pandas/io/rdata/_rdata.pyx

```diff
@@ -0,0 +1,437 @@
+
+    return 0
+
+cdef int handle_dim(
+    const char *name,
+    rdata_type_t dtype,
+    void *data,
+    long count,
+    void *ctx
+) except *:
+    """
+    Parses meta data on non-dataframe objects
+
+    Called once for objects with R dims (matrices, arrays, etc.)).
+    Special condition for character matrices.
+    """
+    lbr = <LibrdataReader>ctx
+
+    cdef int *rdims = <int*>data
+
+    lbr.dims = count
+    lbr.dim_str = {i: rdims[i] for i in range(count)}
+
+    if lbr.rvalues[lbr.tblname]["dtypes"] == {0: "str"}:
+        dim_data = list(lbr.rvalues[lbr.tblname]["data"][0].values())
+
+        n = 0
+        rows, cols = lbr.dim_str.values()
+
+        for col in range(cols):
+            lbr.rvalues[lbr.tblname]["dtypes"][col] = "str"
+            lbr.rvalues[lbr.tblname]["data"][col] = dim_data[n:n+rows]
+            n += rows
+
+    return 0
+
+cdef int handle_column_name(const char *name, int index, void *ctx) except *:
+    """
+    Retrieves column names of data frame
+
+    Returns only non-NULL column names after parsing data.
+    """
+    lbr = <LibrdataReader>ctx
+
+    lbr.colnames[index] = name
+    lbr.rvalues[lbr.tblname]["colnames"] = lbr.colnames
+
+    return 0
+
+cdef int handle_row_name(const char *name, int index, void *ctx) except *:
+    """
+    Retrieves row names of data frame
+
+    Returns only non-NULL row names appear after parsing data.
+    """
+    lbr = <LibrdataReader>ctx
+
+    lbr.rownames[index] = name
+    lbr.rvalues[lbr.tblname]["rownames"] = lbr.rownames
+
+    return 0
+
+cdef int handle_dim_name(const char *name, int index, void *ctx) except *:
+    """
+    Retrieves dim names of matrices or arrays
+
+    Returns only non-NULL dim names appear after parsing data.
+    """
+
+    lbr = <LibrdataReader>ctx
+
+    if (index < lbr.dim_str[0]) and lbr.rownames.get(index) is None:
+        lbr.rownames[index] = name if name != NULL else str(index)
+    else:
+        lbr.rvalues[lbr.tblname]["rownames"] = lbr.rownames
+
+    if index < lbr.dim_str[1]:
+        lbr.colnames[index] = name if name != NULL else str(index)
+    else:
+        lbr.rvalues[lbr.tblname]["colnames"] = lbr.colnames
+
+    return 0
+
+
+class LibrdataParserError(Exception):
+    """
+    Base error class to capture exceptions in librdata parsing.
+    """
+    pass
+
+
+cdef class LibrdataReader:
+    """
+    Base class to read RData files.
+
+    Class interfaces with librdata C library to builds dictionaries
+    of each data frame including data content and meta (dtypes, colnames,
+    and rownames). Callbacks above are used in ``rdata_`` method attributes.
+    """
+    cdef rdata_parser_t *rparser
+    cdef public:
+        int colidx
+        int rows
+        dict rlevels
+        dict rtext
+        bint is_factor
+        dict rownames
+        dict colnames
+        dict rtypes
+        str tblname
+        dict rvalues
+        int dims
+        dict dim_str
+
+    cpdef read_rdata(self, rfile):
```

Comment:
Plain `def` should be fine here too, since this isn't performance sensitive code that needs to be called by Cython.

Reply:
Sounds good.